services/src/main/java/org/keycloak/authentication/requiredactions/SendWelcomeEmail.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.authentication.requiredactions;
+
+import org.keycloak.Config;
+import org.keycloak.authentication.RequiredActionContext;
+import org.keycloak.authentication.RequiredActionFactory;
+import org.keycloak.authentication.RequiredActionProvider;
+import org.keycloak.common.constants.ServiceAccountConstants;
+import org.keycloak.email.EmailException;
+import org.keycloak.email.EmailTemplateProvider;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.services.Urls;
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * A {@link RequiredActionProvider} that sends a welcome email to new users.
+ * <p>
+ * Adds a user attribute {@code "send_welcome_email"} after a sent attempt.
+ * If the {@code "send_welcome_email"} user attribute is present, no welcome email is sent.
+ * </p>
+ *
+ * @author <a href="mailto:thomas.darimont@gmail.com">Thomas Darimont</a>
+ */
+public class SendWelcomeEmail implements RequiredActionFactory, RequiredActionProvider {
+
+    static final String ID = "send_welcome_email";
+
+    @Override
+    public RequiredActionProvider create(KeycloakSession session) {
+        return this;
+    }
+
+    @Override
+    public void init(Config.Scope config) {
+        // NOOP
+    }
+
+    @Override
+    public void postInit(KeycloakSessionFactory factory) {
+        // NOOP
+    }
+
+    @Override
+    public void close() {
+        // NOOP
+    }
+
+    @Override
+    public String getId() {
+        return ID;
+    }
+
+    @Override
+    public String getDisplayText() {
+        return "Send welcome email";
+    }
+
+    @Override
+    public void evaluateTriggers(RequiredActionContext context) {
+
```

Comment:
Is there a way to remove `this` required action after successful mail sending? then I wouldn't need the attribute anymore or at least I could skip the checks entirely :)